### ISSUE TITLE
[API Proposal]: Preprocessing symbol APIs

### ISSUE THREAD

**Issue by Rekkonnect** [dotnet/roslyn#66510]
## Background and Motivation

For the purposes of #66425, we expose a new API called `CreatePreprocessingSymbol`, which creates a preprocessing symbol related to the compilation.

## Proposed API

```diff
 namespace Microsoft.CodeAnalysis
 {
     public abstract class Compilation
     {
+        public IPreprocessingSymbol CreatePreprocessingSymbol(string name);
     }
 }
```

## Usage Example

```csharp
// Given a #region syntax, create a preprocessing symbol named after the region's hint
public static IPreprocessingSymbol CreateFromRegion(SemanticModel model, RegionDirectiveTriviaSyntax regionSyntax)
    => model.Compilation.CreatePreprocessingSymbol(regionSyntax.EndOfDirectiveToken.ToString());
```

## Alternative Designs

- `Compilation.CreatePreprocessingSymbol` might not find immediate usage for the consumers. Currently, `IPreprocessingSymbol` is very rarely used and it doesn't offer much information other than a name and the fact that it could be a C# or a VB preprocessing symbol.

## Risks

There are no breaking changes, and there is no performance regression with the introduction of this API.

**Comment by CyrusNajmabadi** [comment]
I'm not sure I understand the purpose of `public static bool IsIdentifierContainerDirectiveTrivia(SyntaxKind kind);`

**Comment by Rekkonnect** [comment]
It is used to specifically filter through the directive trivia kinds that enable usage of preprocessing symbol names, unlike `#region`, `#pragma` and others. It could also be internalized if its usage isn't deemed important enough for public exposure.

**Comment by CyrusNajmabadi** [comment]
> It is used to specifically filter through the directive trivia kinds that enable usage of preprocessing symbol names

I don't see why a new API is needed for that though. It's just an enumeration of a few kinds, which just a single feature is using.

**Comment by CyrusNajmabadi** [comment]
What does this method do:

```
public PreprocessingSymbolInfo GetPreprocessingSymbolInfo(SyntaxToken token);
```

**Comment by Rekkonnect** [comment]
> What does this method do:
> 
> ```
> public PreprocessingSymbolInfo GetPreprocessingSymbolInfo(SyntaxToken token);
> ```

It's an overload for 
```csharp
public PreprocessingSymbolInfo GetPreprocessingSymbolInfo(SyntaxNode node);
```

**Comment by Rekkonnect** [comment]
> > It is used to specifically filter through the directive trivia kinds that enable usage of preprocessing symbol names
> 
> I don't see why a new API is needed for that though. It's just an enumeration of a few kinds, which just a single feature is using. 

Yeah I'll agree that it's too niche, and can be removed from the proposal.

**Comment by CyrusNajmabadi** [comment]
> It's an overload for

Why is that overload needed?  It would be helpful for the proposal if there were explanations indicating why the existing API is not sufficient. Thanks!

**Comment by Rekkonnect** [comment]
> Why is that overload needed?  It would be helpful for the proposal if there were explanations indicating why the existing API is not sufficient. Thanks!

Right, I had the PR in mind and thought of some context as implciit.

This overload enables seeking the identifier through the SyntaxToken provided by the DefineDirectiveTriviaSyntax and UndefDirectiveTriviaSyntax nodes. This path was chosen over providing specific overloads for these two syntax nodes, and potentially one more for the ConstDirectiveTriviaSyntax for VB (which iirc is desing likewise).

**Comment by CyrusNajmabadi** [comment]
Feels like GetDeclaredSymbol should at least work for #define and #const.

**Comment by Rekkonnect** [comment]
In #66425 it is implemented such that `GetDeclaredSymbol` works for
- `#define`, `#undef`, `#Const`
- `#if`, `#elif`, `#If`, `#ElseIf`

The latter being because the symbol is not "declared" in any of the places, according to my understanding, it is just defined to have the said value, and used in the conditions. It could however be altered to be returned from `GetSymbolInfo` and avoid introducing new behavior on `GetDeclaredSymbol`.

Additionally, the overload
```csharp
public PreprocessingSymbolInfo GetPreprocessingSymbolInfo(SyntaxToken token);
```
could be removed in favor of adding an overload for the `Define`, `Undef` and `Const` directive trivia syntax nodes, which directly refer to the preprocessing symbol info. The question for that is which state of the symbol is returned, the one before the directive trivia, or the one after?

**Comment by jcouv** [comment]
Tagging @AlekseyTs and @333fred as [area owners](https://github.com/dotnet/roslyn/blob/main/docs/area-owners.md#areas) for area api-SemanticModel

**Comment by Rekkonnect** [comment]
Still open question @jcouv @AlekseyTs @333fred do you have any thoughts for moving this forward?

**Comment by jcouv** [comment]
@Rekkonnect We're focused on closing 17.8 (C# 12) for the next two months, so we'll likely not get a design decision on this until then. Please ping again if you don't hear from us by October. Sorry for the wait

**Comment by 333fred** [comment]
For myself, I had been waiting until @CyrusNajmabadi got to the point of being satisfied with the PR. The initial proposal was significantly wider, and it was unclear that was the desired shape. If Cyrus is satisfied with the shape the PR has ended up in, then we can consider this in the next API review.

**Comment by CyrusNajmabadi** [comment]
@Rekkonnect is https://github.com/dotnet/roslyn/pull/66425 updated to only reference this minimal set of apis you're proposing?  i'd liek to see how they are sufficient/necessary for the find-refs operation.  Thanks!

**Comment by jaredpar** [comment]
Moving to backlog due to inactivity

**Comment by Rekkonnect** [comment]
Updated the API proposal to only include the one API that was added. #66425 only adds that method and uses it 
CC @CyrusNajmabadi @jaredpar @jcouv @333fred

**Comment by 333fred** [comment]
### API Review

* No issues with the proposed API.

**Conclusion**: Approved.

**Comment by Rekkonnect** [comment]
Closed via #66425
